### ISSUE TITLE
[feature request] fixing vwii video issues

### ISSUE THREAD

**Issue by Reual4412** [GaryOderNichts/evwii#1]
What we know is that there seems to be a color space issue.
https://gbatemp.net/threads/fixing-vwii-mode.573265/post-9391136
(he says it's his display. But, I suspect that's the output from the cable, as it's even repeatable on my monitor. )

What the issue is specifically is the cable is giving a rec 709 over hdmi, on vwii mode, which is HD color space. Not the correct 609 color space, which is for SD. Which given that the wiiu (in wiiu mode) doesn't properly send RGB range in the first place. (yet somehow does in vwii) It's unlikely Nintendo properly implemented that in the first place.

The second issue, seems to be what's causing the color bleed.
https://gbatemp.net/threads/fixing-vwii-mode.573265/post-9512768

The OG wii does a RGB color space (internally) and then does a conversion to YcBCr, which according to the user, is 4:2:2.
Basically, Nintendo's DMCU? (I'm not sure if that's the exact cause, but considering that it's responsible for VI in vwii.)
instead of doing the correct 4:2:2 format. It's doing  YCrCb 4:4:4. this creates the color bleed and other issues. (Or rather, making them far worse. as the original Wii already had light color bleed.)

Regardless if this is possible to fix or not. Thanks, honestly. It's already amazing to be finally done with the overcropping issue.

**Comment by GaryOderNichts** [comment]
Thanks for mentioning this, I completely missed those posts while skimming over the "Fixing vWii Mode" thread.

Regarding the first issue:
Color Matrix Transformation is disabled in vWii mode, which might confirm the suspicion in the original thread:
> On vWii I believe the issue is that the game is rendering SD 601 colours, but vWii is neglecting to convert them to HD 709 colours before outputting them as 709 colours over HDMI.

Not too sure about the second issue though, the pixel encoding format is set to YCbCr 4:2:2 while in vWii mode.

**Comment by Reual4412** [comment]
> Not too sure about the second issue though, the pixel encoding format is set to YCbCr 4:2:2 while in vWii mode.
>>
Interesting, unless the view port fix fixed the colorbleed issue. (that or it's just purely caused by that color space issue, or something else entirely. One step at a time I suppose)
Also awesome to hear I was able to help somewhat.

**Comment by aaron-clovsky** [comment]
The chroma bleed issue is unfortunately not fixed by evwii for now (it would be amazing if it could be fixed). I have tested both Wind Waker (via Nintendont) and Skyward Sword at 1080p and 480p via HDMI as well as 480p via component while using evwii 0.2. The issue is most clearly visible in skyward sword when looking at the white outline/glow around the goddess statue in the intro - the glow on the right side of the statue is clearly non-white due to the bleeding. For Wind Waker I set evwii to 640x480 to try and reduce scaling artifacts, but bleed could still be seen on Link's right side during the playable intro segments. It honestly looks like a 1-2 pixel horizontal chroma/luma offset (though how that sort of bug is even possible I can't imagine). I can post uncompressed screen captures if that would be helpful.

**Comment by aaron-clovsky** [comment]
For future reference here are comparison captures, both were taken with the following fixes applied to try and get as raw a framebuffer capture as possible:

- 480p fix
- Deflicker Filter disabled (extended)
- Video width: Framebuffer

All captures are fully uncompressed (rgb24) taken with a Magewell capture card and saved to a PNG with ffmpeg.

First capture is from the Wii and was captured unscaled with an OSSC connected with component cables using optimal timings.

![image](https://github.com/GaryOderNichts/evwii/assets/17038276/b22d59c5-75e2-4b79-8808-0307d54b4316)

Second capture is from the Wii U in vWii mode outputting 480p with evwii set to 720x480 and was captured directly via HDMI.

![image](https://github.com/GaryOderNichts/evwii/assets/17038276/8fb67056-29d1-49df-8965-450d2885bb41)

For dramatic effect here's the zoom in (scaled nearest neighbor):

Wii:

![image](https://github.com/GaryOderNichts/evwii/assets/17038276/493e4b7c-a8f7-453f-a4fe-0fdaed89441e)

Wii U (evwii):

![image](https://github.com/GaryOderNichts/evwii/assets/17038276/33b9331f-5543-4417-82a9-1f2d9e3864f2)

Hopefully this helps!

**Comment by Mii-U** [comment]
According to NoobletCheese, the chroma bug seems to be coming from the a misalingment of the CbCr channels, and that it could be solved by outputting RGB instead of YCbCr.

Could a toggle for RGB/YCbCr be added to evWii?

Original post: https://gbatemp.net/threads/fixing-vwii-mode.573265/post-9758960

**Comment by Ingunar** [comment]
**I used Nintendont to force different video modes, and I can confirm this behavior:**
*Just a reminder, the chroma is usually slightly shifted to the left and it is yellowish.*

- Running PAL games on PAL WiiU would not cause bleeding to the red
- Running NTSC games on PAL WiiU would cause bleeding with the red-shifted to the right

The red shift would stay in the HBC and SM, so there would be no more yellowish bleed.
(Shift to yellowish does reset after the reboot)

- The thing that I had a 100% chance to reset the red bleed is setting `TV Type` to `60Hz (480i)` and then changing back to `EDTV/HDTV (480p)`
Most of the time, only changing to `60Hz (480i)` would be enough to fix it.

*When booting the game, staying at 50Hz is the goal:*
- Forcing NTSC, MPAL, or 60Hz on PAL game also results in the red bleed
- Forcing the NTSC game in 50Hz mode would not cause the red bleed
BUT YOU MUST NOT USE Gamecube IPL, and @GaryOderNichts knows why.
*The `__VIInit` reasons*

**Comment by Rulumi** [comment]
I think I already said about this a bit on the linked threads when testing the evwii first version output, but the chroma shift seems to vary between a very ugly one and a fine-ish one according to how the software sets the video output and the VI (and if it reinitializes it or not too?). I imagine this is a bug in Nintendo's implementation on the Wii U as of course doesn't happen on GameCube and Wii. 

If you output on 1:1 without crop and look carefully, when there's more chroma shift, the most horizontal pixels of the vWii output are green on each side. It also always crops one pixel on the game image right side when there's more chroma shift. There are some images that show that below (all with evwii for the Wii U).

Wii U (most chroma shift):
![amarec(20230825-1724) avi_snapshot_00 15_ 2023 08 25_18 05 20](https://github.com/GaryOderNichts/evwii/assets/27870718/7a8d09d9-dd80-45fa-b850-bbdae622acec)

Wii U (less chroma shift):
![amarec(20230825-1731) avi_snapshot_00 14_ 2023 08 25_18 05 14](https://github.com/GaryOderNichts/evwii/assets/27870718/5ad0fafc-d5e4-4aec-9d13-8d157688a0fd)

Wii (2006 unit):
![amarec(20230825-1747) avi_snapshot_00 25_ 2023 08 25_18 05 49](https://github.com/GaryOderNichts/evwii/assets/27870718/dcf653be-4ca1-4b00-a78e-b83522690a16)

**### 300% zoom**

Wii U (most chroma shift):
![image](https://github.com/GaryOderNichts/evwii/assets/27870718/e5449e3e-a8ff-4abb-b6b9-acc09e751683)

Wii U (less chroma shift):
![image](https://github.com/GaryOderNichts/evwii/assets/27870718/e09079ef-cedd-4617-b0e9-2d5d28e3f876)

Wii (2006 unit):
![image](https://github.com/GaryOderNichts/evwii/assets/27870718/eb79c0dc-21ea-4226-a10b-357756e4e84e)

**Comment by MDHEXT** [comment]
I think this would require a rewrite of the DMCU firmware to fix. And I don't believe anyone knows in enough real great detail how the DMCU firmware works exactly to be able to rewrite it to fix all of this.

**Comment by GaryOderNichts** [comment]
> I think this would require a rewrite of the DMCU firmware to fix. And I don't believe anyone knows in enough real great detail how the DMCU firmware works exactly to be able to rewrite it to fix all of this.

I have the DMCU firmware mostly reverse engineered and partially reimplemented. Unfortunately that doesn't automatically help with fixing the chroma shift issues.
There's a lot more testing required to figure out what exactly causes it.

**Comment by Cynirex** [comment]
Hopefully this is something that can be solved in the near future, seeing as DMCU is mostly reverse engineered. Nintendo really dropped the ball with the Wii U.

**Comment by mariomadproductions** [comment]
Is latency also something that people have reproduced/tested? I've noticed it, although I haven't tested it, and its been [reported](https://www.reddit.com/r/SSBM/comments/v4vh1f/comment/ib6kq5q/) as about 20ms, which is pretty high. I'm using a European Wii U, if that makes a difference.

**Comment by wizardkitty** [comment]
If the chroma shift is fixed it would make the wii u almost perfect for most wii games. I would LOVE to see a fix for the issue in the future.

**Comment by nitrate92** [comment]
This is something I haven't been able to reproduce and I'm still not sure how it happened. Please see my GBATemp post here: https://gbatemp.net/threads/wii-u-vwii-output-with-decent-colors.673394/

**Comment by Ingunar** [comment]
Are you thinking about the HDMI 16-235 colour limit?

That is not related to colour bleed.

**Comment by nitrate92** [comment]
I suppose it's not entirely related, but this is another known problem with the vWii's video output. Is there a known fix for that I'm just totally unaware of?

**Comment by Ingunar** [comment]
Based on the info I got, the reason is not USBLGX, but Nintendont or loading of the GC BIOS.
What version of Nintendont are you using, and what is the region of the GC BIOS and game loaded?

**Comment by Fish1717** [comment]
Has there been any progress on addressing the poorly implemented DSMU and these chroma shift issues? The more I learn about Nintendo's work the more I'm dissapointed lol.

**Comment by MDHEXT** [comment]
> Has there been any progress on addressing the poorly implemented DSMU and these chroma shift issues? The more I learn about Nintendo's work the more I'm dissapointed lol.

It's DMCU and yes, a reimplementation is currently being worked on. Though, until that work is completed, modifications to the firmware to improve functionality cannot be made. If you'd like to help or contribute, openDMCUfirm is the name of the project. You should be able to find it on Gary's GitHub profile.

**Comment by Fish1717** [comment]
> > Has there been any progress on addressing the poorly implemented DSMU and these chroma shift issues? The more I learn about Nintendo's work the more I'm dissapointed lol.
> 
> It's DMCU and yes, a reimplementation is currently being worked on. Though, until that work is completed, modifications to the firmware to improve functionality cannot be made. If you'd like to help or contribute, openDMCUfirm is the name of the project. You should be able to find it on Gary's GitHub profile.

Found the project thanks! I’ll definitely contribute thanks!

**Comment by nitrate92** [comment]
> Based on the info I got, the reason is not USBLGX, but Nintendont or loading of the GC BIOS. What version of Nintendont are you using, and what is the region of the GC BIOS and game loaded?

Apologies for the months late reply,

I don't remember what version of Nintendont I was using at the time, as I've since updated. The BIOS and game region were USA. An odd behavior I discovered is that playing with the video modes in vWii will correct the color issues.

https://github.com/user-attachments/assets/6f5b1a1e-3be6-4ab8-8a73-85f8e5feed9e

**Comment by GaryOderNichts** [comment]
> It's DMCU and yes, a reimplementation is currently being worked on. Though, until that work is completed, modifications to the firmware to improve functionality cannot be made.

OpenDMCUFirm is already able to replace the original DMCU firm, making modifications straight-foward. However, it currently doesn't add any improvements over the original firmware. In order to add improvements it is first necessary to figure out what can be improved upon. In the worst case scenario these issues are hardware limitations which cannot be fixed with modifying the DMCU firmware.

> An odd behavior I discovered is that playing with the video modes in vWii will correct the color issues.

Hmm that is indeed interesting, makes it sound like this could be fixed with software. Will take a look at it.

**Comment by Fish1717** [comment]
> > It's DMCU and yes, a reimplementation is currently being worked on. Though, until that work is completed, modifications to the firmware to improve functionality cannot be made.
> 
> OpenDMCUFirm is already able to replace the original DMCU firm, making modifications straight-foward. However, it currently doesn't add any improvements over the original firmware. In order to add improvements it is first necessary to figure out what can be improved upon. In the worst case scenario these issues are hardware limitations which cannot be fixed with modifying the DMCU firmware.
> 
> > An odd behavior I discovered is that playing with the video modes in vWii will correct the color issues.
> 
> Hmm that is indeed interesting, makes it sound like this could be fixed with software. Will take a look at it.

Gary that's great news! It would be amazing if you could fix this color issue with your OpenDMCUFirm and software.  

This kind of fix could eliminate my need for a Wii and even an GameCube console entirely! Please keep us posted :)

**Comment by nitrate92** [comment]
> > It's DMCU and yes, a reimplementation is currently being worked on. Though, until that work is completed, modifications to the firmware to improve functionality cannot be made.
> 
> OpenDMCUFirm is already able to replace the original DMCU firm, making modifications straight-foward. However, it currently doesn't add any improvements over the original firmware. In order to add improvements it is first necessary to figure out what can be improved upon. In the worst case scenario these issues are hardware limitations which cannot be fixed with modifying the DMCU firmware.
> 
> > An odd behavior I discovered is that playing with the video modes in vWii will correct the color issues.
> 
> Hmm that is indeed interesting, makes it sound like this could be fixed with software. Will take a look at it.

Here is some additional context with screenshots I took after switching video modes, if it helps.

<img width="1920" height="1080" alt="Image" src="https://github.com/user-attachments/assets/de338894-500d-4ed2-83cc-46576ac4e80c" />

<img width="1920" height="1080" alt="Image" src="https://github.com/user-attachments/assets/240adeca-7aa6-42af-b1b4-66bd7cf8c395" />

<img width="1920" height="1080" alt="Image" src="https://github.com/user-attachments/assets/89fed675-fe70-4e83-a730-ea6001c8d14c" />

**Comment by Ingunar** [comment]
> > Based on the info I got, the reason is not USBLGX, but Nintendont or loading of the GC BIOS. What version of Nintendont are you using, and what is the region of the GC BIOS and game loaded?
> 
> Apologies for the months late reply,
> 
> I don't remember what version of Nintendont I was using at the time, as I've since updated. The BIOS and game region were USA. An odd behavior I discovered is that playing with the video modes in vWii will correct the color issues.
> 8mb.video-d3v-UzXKhEha.mp4

I can't replicate this.
1) What output are you using 1080p?
2) What plugins are enabled?
3) What is your OS and CFW version?
4) What plugins do you have enabled for vwii mode?

**Comment by GaryOderNichts** [comment]
@nitrate92 Are you sure this isn't an issue with your capture device / TV? The colors never look that gray-ish on my end.

<img width="1920" height="1080" alt="Image" src="https://github.com/user-attachments/assets/f638141c-5b31-4da0-bf78-746dfc2ff4e0" />

Maybe your device picks up the correct range after losing the signal between switching video modes. Feel free to join https://discord.gg/geY4G2NZK9 if you want to troubleshoot this further.

**Comment by Fish1717** [comment]
Good Morning Gary! Thank you again for completing the OpenDMCUFirm project. 

I’m hoping with that project complete that we can use that information to try to fix the offset chroma issue. 

Has anyone identified the root cause on why it’s occurring to try to implement a change with customer firmware?
